### PR TITLE
Api4

### LINKED ISSUE #66: Support API4
It's the wave of the future!

### PR DESCRIPTION
**Pull request by ejegg**
Add support for proxying APIv4 requests. The first patch involves some copy-pasting of API3 code, and the second patch extracts some common functions.  If this looks like the right approach I can add a patch to update the documentation.

Fixes #66

### CODE REVIEW

**Comment by ejegg** [comment]
@bjendres Does this look like a decent approach to you? Let me know if it needs any changes. We're planning to move forward with this locally, and we would love to avoid running a fork.

**Comment by bjendres** [comment]
> @bjendres Does this look like a decent approach to you?

On a breif look: yes. Let my colleagues have a look at it, and we'll merge if we don't see any backward compatibility issues.

> We're planning to move forward with this locally, and we would love to avoid running a fork.

Sure, I understand. It'd be good to have this feature anyway.

**Review by jensschuppe — CHANGES_REQUESTED** [review]
With CiviCRM's *AuthX* REST endpoints, authentication methods are configurable, but *CiviProxy* makes assumptions on which method to use when talking to CiviCRM. This should be either documented or configurable.

Also, we might want to provide all those authentication methods available for the *CiviProxy* endpoint (at least for APIv4), so that clients can connect to *CiviProxy* as if they would to a *real* CivICRM REST endpoint. At least that's what was the contract of *CiviProxy* for the old APIv3 endpoint (which had only one authentication method).

**Comment by ejegg** [comment]
Thanks for all the feedback, @jensschuppe ! I'll try to make those changes soon.

File: proxy/proxy.php
Comment (ejegg) [reply proxy/proxy.php]: OK, latest version has configurable authx flow and can send the site_key guard.

File: proxy/proxy.php
Comment (ejegg) [reply proxy/proxy.php]: Latest version can send site_key guard and api_key via header, xheader, param, and legacyrest flows.

File: proxy/config.dist.php
Comment (jensschuppe) [inline proxy/config.dist.php:110]: Without having reviewed thoroughly, *AuthX* is not bound to APIv4, but allows authentication in general, also for *AuthX*'s APIv3 REST endpoint. I suppose your PR does not cover *AuthX* + APIv3 - that's basically ok, just wanted to note that …

**Comment by agileware-justin** [comment]
Just a suggestion. Consider using Guzzle instead of Curl.

Curl has been slowly replaced with Guzzle throughout the CiviCRM codebase.

**Comment by bjendres** [comment]
> Just a suggestion. Consider using Guzzle instead of Curl.

Understandable, however we're trying to keep the dependencies to an absolute minimum, both for securlty and and compatibility reasons. Remember, it's supposed to be a minimalistic system to be dropped on any (trusted) webspace.

Or is there security concerns about curl that I don't know about?

In any case, please create a separate ticket if you would like further discuss the proposal.

File: proxy/config.dist.php
Comment (ejegg) [reply proxy/config.dist.php:110]: Yep, it would be great to get AuthX support for APIv3 as well, and shouldn't take too much more work. Let's file a separate issue for that and solve it in another PR.

Comment (ejegg) [reply proxy/config.dist.php:110]: I suppose I should rename the settings to simply authx_internal_flow and authx_external_flow, without including the API version.

**Comment by ejegg** [comment]
Latest version of commit changes name of authx config settings in anticipation of being able to use them for APIv3 in the future.

**Comment by ejegg** [comment]
Hi @jensschuppe and @bjendres, I hope you have had a good holiday season. I think I have implemented all of the changes requested in the last review. Is there any other change needed before merging this request?

**Comment by eileenmcnaughton** [comment]
@jensschuppe  - it looks like your feedback has been addressed?

We have this running in production

**Comment by eileenmcnaughton** [comment]
thanks @jensschuppe

**Comment by jensschuppe** [comment]
I discussed with @bjendres and we'd like some more people with an eye on security review this before merging, as the extension is all about security. Also, I opened #71 as a follow-up for supporting APIv3 with *AuthX*.

@bjendres feel free to add to my comment …

**Comment by eileenmcnaughton** [comment]
@bjendres @jensschuppe just reminding you we discussed the joys of pressing that green button on this PR :-)

**Comment by jensschuppe** [comment]
We trust you guys :wink: - released with version `0.7.0`.

**Comment by eileenmcnaughton** [comment]
:-)

**Comment by ejegg** [comment]
Thanks @jensschuppe !

**Pull request merged**
